crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -220,6 +220,64 @@ impl Format<PyFormatContext<'_>> for MaybeParenthesizeExpression<'_> {
                     }
                 }
             },
+            OptionalParentheses::BestFit => match parenthesize {
+                Parenthesize::IfBreaksOrIfRequired => {
+                    parenthesize_if_expands(&expression.format().with_options(Parentheses::Never))
+                        .fmt(f)
+                }
+
+                Parenthesize::Optional | Parenthesize::IfRequired => {
+                    expression.format().with_options(Parentheses::Never).fmt(f)
+                }
+                Parenthesize::IfBreaks => {
+                    let group_id = f.group_id("optional_parentheses");
+                    let f = &mut WithNodeLevel::new(NodeLevel::Expression(Some(group_id)), f);
+                    let mut format_expression = expression
+                        .format()
+                        .with_options(Parentheses::Never)
+                        .memoized();
+
+                    // Don't use best fitting if it is known that the expression can never fit
+                    if format_expression.inspect(f)?.will_break() {
```

Comment:
Ah okay, so this looks for newlines in the IR, interesting. How does `will_break` deal with things that might break conditionally?

Reply:
It looks for things that make group expand for certain: Like hard line breaks or `expand_parent`. The logic ignores any conditions (it returns `true` if content in `if_group_breaks` or `if_group_fits` contains a hard line break). It showed that conditions can mostly be ignored, especially because it is uncommon or even unneeded to have hard line breaks or expand parent inside of conditional content.